### PR TITLE
Add scale kuttl tests

### PR DESCRIPTION
**Pull request by jamepark4**

### CODE REVIEW

File: config/samples/placement_v1beta1_placementapi.yaml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 apiVersion: placement.openstack.org/v1beta1
 kind: PlacementAPI
+namespace: openstack
```
Comment (jamepark4) [inline config/samples/placement_v1beta1_placementapi.yaml:3]: Would it make more sense to create a config/samples/kuttl_placement_v1beta1_placementapi.yaml file and use that for deploying placement versus leveraging this file?

File: kuttl-test.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#
+# EXECUTION (from install_yamls repo root):
+#
+#   make kuttl_keystone
```
Comment (EliadCohen) [inline kuttl-test.yaml:4]: Keystone? (nit)

Comment (jamepark4) [reply kuttl-test.yaml:4]: Thanks!

File: kuttl-test.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#
+# 1. Latest version of kuttl is installed at /usr/local/bin/kubectl-kuttl
+#    - wget https://github.com/kudobuilder/kuttl/releases/download/v0.11.1/kubectl-kuttl_0.11.1_linux_x86_64
+#    - mv kubectl-kuttl_0.11.1_linux_x86_64 /usr/local/bin/kubectl-kuttl
+#    - chmod 755 /usr/local/bin/kubectl-kuttl
+# 2. An OCP 4.10+ CRC cluster with Podified Operators has been deployed
+# 3. CLI user has access to $KUBECONFIG
+# 4. The environment variable INSTALL_YAMLS is set to the the path of the
+#    install_yamls repo
+
```
Comment (gibizer) [inline kuttl-test.yaml:19]: I tried to follow this instructions:
I did the following against a fresh crc:
// in install_yamls
make keystone_kuttl

But the target failed with:
```
 logger.go:42: 19:13:20 | change_keystone_config/4-change_keystone_config | starting test step 4-change_keystone_config
    logger.go:42: 19:13:20 | change_keystone_config/4-change_keystone_config | running command: [sh -c oc patch keystoneapi -n openstack keystone --type='json' -p='[{"op": "replace", "path": "/spec/customServiceConfig", "value": "[DEFAULT]\ndebug = false"}]'
        ]
    logger.go:42: 19:13:20 | change_keystone_config/4-change_keystone_config | keystoneapi.keystone.openstack.org/keystone patched
    logger.go:42: 19:13:23 | change_keystone_config/4-change_keystone_config | running command: [sh -c $KEYSTONE_KUTTL_DIR/../common/scripts/check_debug_in_keystone_pod_logs.sh --reverse]
    logger.go:42: 19:13:23 | change_keystone_config/4-change_keystone_config | /home/gibi/upstream/git/openstack-k8s-operators/install_yamls/out/operator/keystone-operator/tests/kuttl/tests/../common/scripts/check_debug_in_keystone_pod_logs.sh: 5: [: --reverse: unexpected operator
    logger.go:42: 19:13:24 | change_keystone_config/4-change_keystone_config | error: arguments in resource/name form may not have more than one slash
    logger.go:42: 19:13:25 | change_keystone_config/4-change_keystone_config | running command: [sh -c $KEYSTONE_KUTTL_DIR/../common/scripts/check_debug_in_keystone_pod_logs.sh --reverse]
    logger.go:42: 19:13:25 | change_keystone_config/4-change_keystone_config | /home/gibi/upstream/git/openstack-k8s-operators/install_yamls/out/operator/keystone-operator/tests/kuttl/tests/../common/scripts/check_debug_in_keystone_pod_logs.sh: 5: [: --reverse: unexpected operator
```

Did I missed something?

**Review by gibizer — COMMENTED** [review]
I tried to follow the instructions to run the test but I failed on the keystone prerequisites.

File: config/samples/placement_v1beta1_placementapi.yaml (hunk 1 above)
Comment (jamepark4) [reply config/samples/placement_v1beta1_placementapi.yaml:3]: Yea it does not need anything special at this time so I think I would prefer the option to patch the namespace with kustomize.

File: kuttl-test.yaml (hunk 3 above)
Comment (jamepark4) [reply kuttl-test.yaml:29]: Can you retry with the updated steps whenever you have time?

File: kuttl-test.yaml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+#    2a. CPUS=12 MEMORY=12016 make crc
+#    2b. oc login -u kubeadmin -p 12345678 https://api.crc.testing:6443
+#    2c. eval $(crc oc-env)
+#    2d. make crc_storage
+#    2e. make input
+#    2f. make keystone
+#    2g. make mariadb
+#    2h. make keystone_deploy
+#    2i. make mariadb_deploy
+# 3. The above environment variables PLACEMENT_KUTTL_CONF and PLACEMENT_KUTTL_TEST_DIR are set
```
Comment (gibizer) [inline kuttl-test.yaml:28]: Is it PLACEMENT_KUTTL_TEST_DIR or PLACEMENT_KUTTL_DIR? I think the test code uses the latter.

File: kuttl-test.yaml (hunk 3 above)
Comment (gibizer) [reply kuttl-test.yaml:29]: Sorry I mixed up things yesterday with the keystone_kuttl target. 

Today I followed the new instructions (thanks for that!). 

```
gibi@bedrock:~/upstream/git/openstack-k8s-operators/install_yamls                                                                                                                       $ env | grep PLACEMENT
PLACEMENT_KUTTL_CONF=../placement-operator/kuttl-test.yaml
PLACEMENT_KUTTL_DIR=../placement-operator/test/kuttl/tests
$ kubectl-kuttl test --config ${PLACEMENT_KUTTL_CONF} ${PLACEMENT_KUTTL_DIR} 
[...snip...]
    logger.go:42: 17:07:29 | placement_scale/1-deploy_placement | PlacementAPI:openstack/placement created
    logger.go:42: 17:07:29 | placement_scale/1-deploy_placement | test step completed 1-deploy_placement
    logger.go:42: 17:07:29 | placement_scale/2-patch_placement_deploy | starting test step 2-patch_placement_deploy
    logger.go:42: 17:07:29 | placement_scale/2-patch_placement_deploy | running command: [sh -c oc patch placementapi -n openstack placement --type='json' -p='[{"op": "replace", "path": "/spec/secret", "value":"osp-secret"}]'
        oc patch placementapi -n openstack placement --type='json' -p='[{"op": "replace", "path": "/spec/storageClass", "value":"local-storage"}]'
        ]
    logger.go:42: 17:07:29 | placement_scale/2-patch_placement_deploy | placementapi.placement.openstack.org/placement patched
    logger.go:42: 17:07:29 | placement_scale/2-patch_placement_deploy | Warning: unknown field "spec.storageClass"
    logger.go:42: 17:07:29 | placement_scale/2-patch_placement_deploy | placementapi.placement.openstack.org/placement patched (no change)
    logger.go:42: 17:07:33 | placement_scale/2-patch_placement_deploy | running command: [sh -c source $PLACEMENT_KUTTL_DIR/../lib/helper_functions.sh
        apiEndpoints=$(oc get -n openstack PlacementAPI placement -o go-template-file=$PLACEMENT_KUTTL_DIR/../go_templates/apiEndpoints.gotemplate)
        assert_regex $apiEndpoints 'http:\/\/placement-internal\.openstack\.svc.*'
        assert_regex $apiEndpoints 'http:\/\/placement-public-openstack\.apps.*'
        ]
    logger.go:42: 17:07:33 | placement_scale/2-patch_placement_deploy | sh: 1: source: not found
    logger.go:42: 17:07:33 | placement_scale/2-patch_placement_deploy | sh: 3: assert_regex: not found
    logger.go:42: 17:07:33 | placement_scale/2-patch_placement_deploy | sh: 4: assert_regex: not found
    logger.go:42: 17:07:34 | placement_scale/2-patch_placement_deploy | running command: [sh -c source $PLACEMENT_KUTTL_DIR/../lib/helper_functions.sh
        apiEndpoints=$(oc get -n openstack PlacementAPI placement -o go-template-file=$PLACEMENT_KUTTL_DIR/../go_templates/apiEndpoints.gotemplate)
        assert_regex $apiEndpoints 'http:\/\/placement-internal\.openstack\.svc.*'
        assert_regex $apiEndpoints 'http:\/\/placement-public-openstack\.apps.*'
        ]
    logger.go:42: 17:07:34 | placement_scale/2-patch_placement_deploy | sh: 1: source: not found
    logger.go:42: 17:07:34 | placement_scale/2-patch_placement_deploy | sh: 3: assert_regex: not found
    logger.go:42: 17:07:34 | placement_scale/2-patch_placement_deploy | sh: 4: assert_regex: not found

```

What seems to be the root cause is that `sh -c source` returns `sh: 1: source: not found`.  
In my env:
```
$ which sh
/usr/bin/sh
$ ll /usr/bin/sh
lrwxrwxrwx 1 root root 4 Jan  5 14:20 /usr/bin/sh -> dash
$ 
$ sh -c source 
sh: 1: source: not found
```
I know we hit this before that the current test code depends on bash but kuttl uses sh -c to run the script. However I don't remember how I solved it locally.

Comment (jamepark4) [reply kuttl-test.yaml:29]: I updated it to use . which is suppose to be more universal.  Not seeing any issues on my side when I try this, let me know if this works in yours.

File: kuttl-test.yaml (hunk 4 above)
Comment (jamepark4) [reply kuttl-test.yaml:28]: Thanks I updated the env var in the instructions to be more descriptive and forgot to update the assert yaml, it is PLACEMENT_KUTTL_TEST_DIR.

File: kuttl-test.yaml (hunk 3 above)
Comment (gibizer) [reply kuttl-test.yaml:29]: While `.` is POSIX shell compatible (compared to `source` that is a bashism) the code that is loaded by `.`  in the test case only compatible with bash (e.g. `[[` is bashism). So if the `sh` points to other than bash then the loaded code from lib will fail on incompatibility.

I found an ugly but safe workaround for my env. I can modify the PATH for the kuttl execution to include a local directory where `sh` is soft linked to `/bin/bash`. This way I don't have to alter my local system but can still force kuttl to run /bin/bash as `sh`. So the tests are passing now for me too:
```
$ env | grep PLACEMENT
PLACEMENT_KUTTL_CONF=../placement-operator/kuttl-test.yaml
PLACEMENT_KUTTL_TEST_DIR=../placement-operator/test/kuttl/tests
$ ll /home/gibi/bin/sh-to-bash
total 0
lrwxrwxrwx 1 gibi gibi 9 Apr 13 09:02 sh -> /bin/bash
$ PATH="/home/gibi/bin/sh-to-bash:$PATH" kubectl-kuttl test --config ${PLACEMENT_KUTTL_CONF} ${PLACEMENT_KUTTL_TEST_DIR} 
[...snip...]
    harness.go:399: run tests finished
    harness.go:508: cleaning up
    harness.go:563: removing temp folder: ""
--- PASS: kuttl (66.72s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/placement_scale (56.66s)
PASS
```

File: kuttl-test.yaml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+# and install_yamls)
+#    2a. CPUS=12 MEMORY=12016 make crc
+#    2b. oc login -u kubeadmin -p 12345678 https://api.crc.testing:6443
+#    2c. eval $(crc oc-env)
+#    2d. make crc_storage
+#    2e. make input
+#    2f. make keystone
+#    2g. make mariadb
+#    2h. make keystone_deploy
+#    2i. make mariadb_deploy
```
Comment (gibizer) [inline kuttl-test.yaml:27]: There is a missing step. We need to deploy the placement-operator somehow as the test only creates PlacementAPI CRDs but that won't be reconciled without placement-operator running. One way to do that is `make placement` in install_yamls but other can be `OPERATOR_TEMPLATES=./templates make run` in the placement-operator repo root.

File: test/kuttl/common/errors_cleanup_placement.yaml [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  labels:
+    public: "true"
+    service: placement
+  namespace: openstack
+spec:
+  port:
+    targetPort: placement-public
+  to:
+    kind: Service
+    name: placement-public
```
Comment (gibizer) [inline test/kuttl/common/errors_cleanup_placement.yaml:91]: We can check that the config and script ConfigMap for placement is also deleted by the placement-operator when PlacementAPI is deleted.

These are the config maps created for PlacementAPI/placement.
```
$ oc get configmap -l placement.openstack.org/name=placement
NAME                    DATA   AGE
placement-config-data   5      2m30s
placement-scripts       2      2m30s
```

File: test/kuttl/common/assert_sample_deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+    targetPort: placement-public
+  to:
+    kind: Service
+    name: placement-public
+status:
+  ingress:
+  - conditions:
+    - status: "True"
+      type: Admitted
+---
```
Comment (gibizer) [inline test/kuttl/common/assert_sample_deployment.yaml:228]: We can also check that the config maps exists. 
```
$ oc get configmap -l placement.openstack.org/name=placement
NAME                    DATA   AGE
placement-config-data   5      2m30s
placement-scripts       2      2m30s
```

As kuttl cannot match a field partially I suggest not to try to match the Data field of the ConfigMap. Maybe later if we want to test reconfiguration then we can use assert_regex to check for parts of the placement-config-data.Data field to see if a custom configuration ended up in the given config map. But keystone actually checks the config within the service Pod in the reconfiguration test. Which also OK for me.

File: test/kuttl/common/patch_placement_deploy.yaml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+apiVersion: kuttl.dev/v1beta1
+kind: TestStep
+commands:
+  - script: |
+      oc patch placementapi -n openstack placement --type='json' -p='[{"op": "replace", "path": "/spec/secret", "value":"osp-secret"}]'
+      oc patch placementapi -n openstack placement --type='json' -p='[{"op": "replace", "path": "/spec/storageClass", "value":"local-storage"}]'
```
Comment (gibizer) [inline test/kuttl/common/patch_placement_deploy.yaml:6]: PlacementAPI.Spec  https://github.com/openstack-k8s-operators/placement-operator/blob/508390fe332e24c97625ca7d4c15088e9c09d158/api/v1beta1/placementapi_types.go#L38 does not have storageClass field. So this line should be removed.

File: test/kuttl/tests/placement_scale/05-assert.yaml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+#
+
+apiVersion: placement.openstack.org/v1beta1
+kind: PlacementAPI
+metadata:
+  finalizers:
+  - PlacementAPI
+  name: placement
+  namespace: openstack
+spec:
```
Comment (gibizer) [inline test/kuttl/tests/placement_scale/05-assert.yaml:15]: you can shrink the assert here to be as limited as in 04-assert.yaml

File: test/kuttl/tests/placement_scale/05-assert.yaml [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  secret: osp-secret
+status:
+  databaseHostname: openstack
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: placement
+spec:
+  replicas: 0
```
Comment (gibizer) [inline test/kuttl/tests/placement_scale/05-assert.yaml:39]: Can we have comments in these yaml files? If so then please add a note that we are asserting spec.replicas here instead for status.availableReplicas as that field is not provided in the output if the value of it is 0.

**Review by gibizer — CHANGES_REQUESTED** [review]
I have couple of small requests inline but overall it looks good and I was able to run the test successfully locally.

+1 for not having any install_yamls dependency within the test cases. (Having it in the preparation steps as examples how to prepare a test env is OK as it is just an example)

+1 for having a separate assert_regex function

File: kuttl-test.yaml (hunk 3 above)
Comment (jamepark4) [reply kuttl-test.yaml:29]: Ack thanks for clarifying what was happening and I'm glad you have a solution

File: kuttl-test.yaml (hunk 5 above)
Comment (jamepark4) [reply kuttl-test.yaml:27]: Ack let me try the latter approach and will update accordingly.

File: test/kuttl/common/errors_cleanup_placement.yaml (hunk 6 above)
Comment (jamepark4) [reply test/kuttl/common/errors_cleanup_placement.yaml:91]: Good idea, will update to check for this as well.

File: test/kuttl/common/patch_placement_deploy.yaml (hunk 8 above)
Comment (jamepark4) [reply test/kuttl/common/patch_placement_deploy.yaml:6]: Will remove, thanks.

File: test/kuttl/tests/placement_scale/05-assert.yaml (hunk 10 above)
Comment (jamepark4) [reply test/kuttl/tests/placement_scale/05-assert.yaml:39]: Yes that shouldn't be a problem to add.

**Review by gibizer — APPROVED** [review]
Looks good to me. Thanks!

**Pull request merged**